File: leetcode/java/12.integer-to-roman.java
/*
 * [12] Integer to Roman
 *
 * https://leetcode.com/problems/integer-to-roman
 *
 * Given an integer, convert it to a roman numeral.
 *
 * Input is guaranteed to be within the range from 1 to 3999.
 */
public class Solution {
    public String intToRoman(int num) {
        int a = num / 1000;
        num = num - a * 1000;
        int b = num / 100;
        num = num - b * 100;
        int c = num / 10;
        num = num - c * 10;
        int d = num;

        String roman = get("M", "", "", a) +
                get("C", "D", "M", b) +
                get("X", "L", "C", c) +
                get("I", "V", "X", d);

        return roman;
    }

    private String get(String one, String five, String ten, int a) {
        if (a < 4) {
            return accu(one, a);
        } else if (a == 4) {
            return one + five;
        } else if (a < 9) {
            return five + accu(one, a - 5);
        } else {
            return one + ten;
        }
    }

    private String accu(String sym, int ct) {
        String res = "";
        for (int i = 0; i < ct; i++) {
            res += sym;
        }
        return res;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.intToRoman(1));
        System.out.println(s.intToRoman(2));
        System.out.println(s.intToRoman(5));
        System.out.println(s.intToRoman(10));
        System.out.println(s.intToRoman(25));
        System.out.println(s.intToRoman(68));
        System.out.println(s.intToRoman(1500));
        System.out.println(s.intToRoman(88));
        System.out.println(s.intToRoman(1800));
    }
}

File: leetcode/javascript/12.integer-to-roman.js
/*
 * [12] Integer to Roman
 *
 * https://leetcode.com/problems/integer-to-roman
 *
 * algorithms
 * Medium (44.91%)
 * Total Accepted:    123.7K
 * Total Submissions: 271.5K
 * Testcase Example:  '1'
 *
 * Given an integer, convert it to a roman numeral.
 * 
 * 
 * Input is guaranteed to be within the range from 1 to 3999.
 *
 * I = 1
 * V = 5
 * X = 10
 * L = 50
 * C = 100
 * D = 500
 * M = 1000
 */

const Romans = [
  null,
  ['I', 'V', 'X'],
  ['X', 'L', 'C'],
  ['C', 'D', 'M'],
  ['M']
];

function repeat(str, n) {
  let res = '';
  for (let i = 0; i < n; i++) {
    res += str;
  }
  return res;
}

function convertDigitToRomain(num, digit) {
  let n = Number(num);
  if (n < 4) {
    return repeat(Romans[digit][0], n);
  } else if (n === 4) {
    return Romans[digit][0] + Romans[digit][1];
  } else if (n < 9) {
    return Romans[digit][1] + repeat(Romans[digit][0], n - 5);
  } else if (n === 9) {
    return Romans[digit][0] + Romans[digit][2];
  }

  return ''; // num = '0'
}

/**
 * @param {number} num
 * @return {string}
 */
var intToRoman = function(num) {
  let str = num.toString();
  let res = '';
  for (let i = str.length; i > 0; i--) {
    res += convertDigitToRomain(str[str.length -  i], i);
  }

  return res;
};
